File: java/src/test/java/io/cucumber/tagexpressions/ParsingTest.java
package io.cucumber.tagexpressions;

import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;
import org.yaml.snakeyaml.Yaml;

import java.io.IOException;
import java.nio.file.Paths;
import java.util.List;
import java.util.Map;

import static java.nio.file.Files.newInputStream;
import static org.junit.jupiter.api.Assertions.assertEquals;

class ParsingTest {

    private static List<Map<String, String>> acceptance_tests_pass() throws IOException {
        return new Yaml().loadAs(newInputStream(Paths.get("..", "testdata", "parsing.yml")), List.class);
    }

    @ParameterizedTest
    @MethodSource
    void acceptance_tests_pass(Map<String, String> expectation) {
        Expression expr = TagExpressionParser.parse(expectation.get("expression"));
        String formatted = expectation.get("formatted");
        assertEquals(formatted, expr.toString());

        Expression expr2 = TagExpressionParser.parse(formatted);
        assertEquals(formatted, expr2.toString());
    }
}

File: php/tests/ParsingTest.php
<?php

namespace Cucumber\TagExpressions\Tests;

use Cucumber\TagExpressions\TagExpressionParser;
use PHPUnit\Framework\Attributes\DataProvider;
use PHPUnit\Framework\TestCase;
use Symfony\Component\Yaml\Yaml;

class ParsingTest extends TestCase
{
    #[DataProvider('provideExpectations')]
    public function testParsing(string $expression, string $formatted): void
    {
        $expr1 = TagExpressionParser::parse($expression);
        self::assertEquals($formatted, (string) $expr1);

        $expr2 = TagExpressionParser::parse($formatted);
        self::assertEquals($formatted, (string) $expr2);
    }

    /**
     * @return iterable<array{string, string}>
     */
    public static function provideExpectations(): iterable
    {
        /** @var list<array{expression: string, formatted: string}> $data */
        $data = Yaml::parseFile(__DIR__ . '/../../testdata/parsing.yml');

        foreach ($data as $item) {
            yield [$item['expression'], $item['formatted']];
        }
    }
}
